jetstream_pt/ray_engine_worker.py

```diff
@@ -0,0 +1,851 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implement Jet Engine API."""
+import queue
+from typing import Any, List, Optional, Tuple, Union
+import threading
+import functools
+import humanize
+
+
+from etils import epath
+import safetensors
+from flax import struct
+import jax
+from jax import numpy as jnp
+from jax.experimental import multihost_utils
+import torch
+import numpy as np
+import ray
+from torch.utils import _pytree as pytree
+import torch_xla2
+
+from jetstream.engine import engine_api, tokenizer_pb2, token_utils
+
+from jetstream_pt.third_party.llama2 import model_exportable, model_args
+
+from jetstream_pt import cache_manager
+from jetstream_pt import quantize
+from jetstream_pt.environment import JetEngineEnvironment, JetEngineEnvironmentData
+
+
+Mesh = jax.sharding.Mesh
+P = jax.sharding.PartitionSpec
+
+Params = jax.Array
+PrefillInputs = np.ndarray
+
+
+@struct.dataclass
+# pylint: disable-next=all
+class Prefix:
+  token: jax.Array  # [1, seqlen]
+  caches: List[Tuple[jax.Array, jax.Array]]
+  seq_len: int  # true seqlen front pad
+
+
+@struct.dataclass
+# pylint: disable-next=all
+class DecodeState:
+  tokens: jax.Array  # [batch_size, seqlen]
+  caches: List[Tuple[jax.Array, jax.Array]]
+  cache_scales: List[
+      Tuple[jax.Array, jax.Array]
+  ]  # only present in quantized kv
+  current_position: int
+  lens: jax.Array  # [batch_size, 1]
+  input_pos: jax.Array  # [batch_size, 1] input pos for each slot
+  mask: jax.Array  # [batch_size, seqlen] -inf for invalid; 0 for valid
+
+
+# NOTE model specific
+
+
+@ray.remote
+# pylint: disable-next=all
+class PyTorchEngineRayWorker:
+  """Wraps functions to the Jet Engine API format."""
```

Comment:
It could be helpful to add more documentation here, for instance:

```
class PyTorchEngineRayWorker:
  """Ray actor representation for a PyTorch engine worker.

  PyTorchEngineRayWorker enables multi-host serving for models that exceed the memory
  capabilities of single-host TPU VMs with Ray. PyTorchEngineRayWorker should be used with a "leader" engine,
  e.g. `PyTorchRayEngine`.

  Note: For `PyTorchRayEngine` to return consistent results, it's important that `PyTorchEngineRayWorker` is able
  to maintain its state within process and only return results once its transferred to CPU device.

  """
```

(as I write this out, if we're renaming the above to `PyTorchRayEngine` then this should be `PyTorchRayWorker`)

Reply:
Agree on renaming, updated.